accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+        rules.AddAccessRule(
+            new FileSystemAccessRule(@"fabrikam\log-writers",
+                                     FileSystemRights.FullControl,
+                                     AccessControlType.Allow)
+        );
+
+        Directory.CreateDirectory(loggingDirectory, rules);
+    }
+
+    return loggingDirectory;
+}
+```
+
+After that, the diagnostic disappears automatically.
+
+## Requirements
+
+### Goals
+
+* Developers get diagnostics when they inadvertently use platform-specific APIs.
+    - All diagnostics are provided on the command line as well as in the IDE
+    - The developer doesn't have to resort to explicit suppressions to indicate
+      intent. Idiomatic gestures, such as guarding the call with a platform
+      check or marking a method as platform-specific, automatically suppress
+      these diagnostics. However, explicit suppressions via the usual means
+      (`NoWarn`, `#pragma`) will also work.
+* The analyzer helps library authors to express their intent so that their
+  consumers can also benefit from these diagnostics.
+    - Library authors can annotate their platform-specific APIs without taking a
+      dependency.
+    - Library consumers don't have to install a NuGet package to get diagnostics
+      when using platform-specific APIs
+
+### Non-Goals
+
+* Modelling partially portable APIs (that is APIs are applicable to more than
+  one platform but no all platforms). Those should Be modeled as capability
+  APIs, as as `RuntimeInformation.IsDynamicCodeSupported` for ref emit. This
+  would be a separate analyzer.
+* Shipping the platform-specific analyzer and/or annotations out-of-band.
+
+## Design
+
+### Attribute
+
+OS bindings and platforms-specific BCL APIs will be annotated using a set of
+custom attributes:
+
+```C#
+namespace System.Runtime.InteropServices
+{
+    public partial struct OSPlatform
+    {
+        // Existing properties
+        // public static OSPlatform FreeBSD { get; }
+        // public static OSPlatform Linux { get; }
+        // public static OSPlatform OSX { get; }
+        // public static OSPlatform Windows { get; }
+        public static OSPlatform Android { get; }
+        public static OSPlatform iOS { get; }
+    }
+
+    public partial static class RuntimeInformation
+    {
+        // Existing API
+        // public static bool IsOSPlatform(OSPlatform osPlatform);
+
+        // Check for the OS with a >= version comparison
+        // Used to guard APIs that were added in the given OS release.
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major);
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor);
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor, int build);
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor, int build, int revision);
+
+        // Allows checking for the OS with a < version comparison
```

Comment:
It’s a `<=` comparison, right? That’s how I interpret “OS platform and earlier” (and I’m kinda wondering why it’s not “OrEarlier”)

Reply:
No it's `<`. The idea is that this API is used to guard APIs that were removed. That's why it's `AndEarlier` to be in constrast to `IsOSPlatfromOrLater`. But I'd lie if I claimed that I'm a 100% in love with those names.

Reply:
Hrm, “and earlier” does not read as an exclusive range to me at all. Saying “supported on iOS 14.0 and earlier” and “supported on iOS 14.0 or earlier” both say `<= 14.0` to me.

Perhaps `IsOSPlatformEarlierThan`?

Reply:
I don't think I have ever done a check to see if it is on an earlier version of the OS or not. You always check if it is the current platform or newer.

https://github.com/jamesmontemagno/Hanselman.Forms/blob/vnext/src/Android/Helpers/Environment.cs#L25-L29

Reply:
I feel like you are checking two thing here, which is a bit complicated.... 

Why not just it be a version check, and pass in a Version instead of ints https://docs.microsoft.com/en-us/dotnet/api/xamarin.essentials.deviceinfo.version?view=xamarin-essentials#Xamarin_Essentials_DeviceInfo_Version 

I would imagine I could just do this:

```
if(DeviceInfo.Platform == DevicePlatform.iOS
    && DeviceInfo.Version.Major >= 13)
{
  /// running iOS 13 stuff to access apis
}
```

Most of the time I would already be in platform specific code so I don't need to check the platform at all and can just check the version.

Reply:
I've never seen a negative application version number. Having all these overloads seems redundant. Semantically these two are identical:
```c#
IsOSPlatformOrLater(OSPlatform.iOS, 14);
```
And
```c#
IsOSPlatformOrLater(OSPlatform.iOS, 14, 0);
```
You could replace them all with the single api:
```c#
public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor = 0, int build = 0, int revision = 0);
```

Reply:
Could also just take in a Version :), but I like the above too. no need for all the overloads, default to 0.

Reply:
> Could also just take in a Version :)

I don't want to accept `Version` for two reasons:

1. It makes the call site more verbose
2. It makes it harder for the analyzer to extract the checked version because people will be inclined to put the version into a local variable before passing it to the `IsPlatformXxx` API.

Reply:
@jamesmontemagno 

> I don't think I have ever done a check to see if it is on an earlier version of the OS or not. You always check if it is the current platform or newer.

Let's say an API `X` was obsoleted in version 12, which also adds the replacement `Y`. In version 13 the API was removed. You need to support 11-13. In this case you can get away with code like this:

```C#
if (RuntimeInformation.IsOSPlatformOrLater(OSPlatform.iOS, 12))
{
    AppleApiY();
}
else
{
    AppleApiX();
}
```

But now let's say the places where you need to call `X` and `Y` are different, which can happen depending on how the APIs was removed. Or take an API that Apple removed because the action performed by it is now done automatically (IOW, there is no `Y`). In this case you'd be forced to write code like this:

```C#
if (RuntimeInformation.IsOSPlatformOrLater(OSPlatform.iOS, 12))
{
    // Now automatic
}
else
{
    AppleApiX();
}
```

Which feels slightly off. It seems more natural to write code like this:

```C#
if (RuntimeInformation.IsOSPlatformEarlierThan(OSPlatform.iOS, 12))
{
    // Before iOS 12, we need to call this API.
    AppleApiX();
}
```

Please note that negating the check of `IsOSPlatformOrLater()` isn't the correct thing because that would make you call `AppleApiX()` on non-iOS platforms. So without the API the expression would be:

```C#
if (RuntimeInformation.IsOSPlatform(OSPlatform.iOS) && !RuntimeInformation.IsOSPlatformOrLater(OSPlatform.iOS, 12))
{
    // Before iOS 12, we need to call this API.
    AppleApiX();
}
```

Reply:
@mconnew 

> You could replace them all with the single api:
> 
> ```cs
> public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor = 0, int build = 0, int revision = 0);
> ```

Normally I'd agree but historically we know that developers get version checks wrong all the time. Hence, I want to make the API shape as simple as possible. We also know that optional parameters aren't intuitive to everyone. And strictly speaking the allow odd calling patterns, like `IsOSPlatformOrLater(OSPlatform.iOS, 12, build: 23)`.